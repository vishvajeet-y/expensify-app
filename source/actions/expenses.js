import { v4 as uuidv4 } from 'uuid'
import database from '../Firebase/firebase'

// export const addExpense=({  //destructurng payload and asssigning default value
//     description='',
//     note='',
//     ammount=0,
//     createAt=0
// }={})=>({
//     type:'ADD_EXPENSE',
//     expense:{
//     id:uuidv4(),
//     description,
//     note,
//     ammount,
//     createAt
//     }
// })

export const addExpense=(expense)=>{
    return {
        type:'ADD_EXPENSE',
        expense
    }
}

export const startAddExpense=(expenseData={})=>{
    return (dispatch)=>{
        const {
            description='',
            note='',
            createAt=0,
            ammount=0
        }=expenseData

        const expense={description,note,ammount,createAt}

      return  database.ref('expenses').push(expense).then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        })
    }
}

//Remove Expense
export const removeExpense=({id}={})=>({
    type:'REMOVE_EXPENSE',
    id
})

//Edit Expense
export const editExpense=(id,updates)=>({
   type:'EDIT_EXPENSE',
   id,updates
})
