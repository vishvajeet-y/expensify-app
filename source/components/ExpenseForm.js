import React from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'


// const now=moment()
// console.log(now.format('MMM Do YYYY'))

export default class ExpenseForm extends React.Component{

    constructor(props){
        super(props);
       this.state={
            description:props.expense?props.expense.description:'',
            note:props.expense?props.expense.note:'',
            ammount:props.expense?props.expense.ammount.toString():'',
            createAt:props.expense?moment(props.expense.createAt):moment(),
            calenderFocused:false,
            error:''
        };
    }

    onDescriptionChange=(e)=>{
              this.setState(()=>({description:e.target.value}))
    }

    onNoteChange=(e)=>{
        this.setState(()=>({note:e.target.value}))
    }

    onAmmountChange=(e)=>{
        const ammount=e.target.value
        if(!ammount||ammount.match(/^\d{1,}(\.\d{0,2})?$/))
        {
            this.setState(()=>({ammount}))
        }
    }
 
     onDateChange=(createAt)=>{
         if(createAt)
         this.setState(()=>({createAt}))
     }

     onFocusChange=({focused})=>{
         this.setState(()=>({calenderFocused:focused}))
     }
     
     onSubmit=(e)=>{
         e.preventDefault()

         if(!this.state.description||!this.state.ammount){
            this.setState(()=>({error:'Please fill both Description and Ammount'}))
         }
         else{
            this.setState(()=>({error:''}))
             this.props.onSubmit1({
                 description:this.state.description,
                 ammount:parseFloat(this.state.ammount,10),
                 createAt:this.state.createAt.valueOf(),
                 note:this.state.note
             })
         }
     }

    render(){
        return  <div>
             {this.state.error&&<p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
            <input type='text' placeholder='Description' autoFocus value={this.state.description} 
               onChange={this.onDescriptionChange}/>
            <input type='text' placeholder='Ammount' value={this.state.ammount}
                onChange={this.onAmmountChange}
                />
                <SingleDatePicker
                date={this.state.createAt}
                onDateChange={this.onDateChange}
                focused={this.state.calenderFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                />
            <textarea placeholder='Add a note for your expense (optional)'value={this.state.note}
            onChange={this.onNoteChange}
            ></textarea>
            <button>Add Expense</button>
            </form>
             </div>
        
    }
}