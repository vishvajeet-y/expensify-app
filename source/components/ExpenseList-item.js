import React from 'react'
import moment from 'moment'
import numeral from 'numeral'

import {Link} from 'react-router-dom'
numeral.register('locale', 'in', {
  delimiters: {
    thousands: ',',
    decimal: '.'
  },
  abbreviations: {
    thousand: 'k',
    million: 'm',
    billion: 'b',
    trillion: 't'
  },
  currency: {
    symbol: 'â‚¹'
  }
});
{numeral.locale('in')}
const ExpenseListItem= ({id,description,ammount,createAt})=>{

  return  <div>
    <Link to={`./edit/${id}`}> <h3>{description}</h3></Link>
    
    <p>
    
    Ammount: {numeral(ammount).format('$0,0.00')} </p>
      
   <p> Date :{moment(createAt).format('Do MMMM,YYYY')}</p>  
    
   
    </div>
    }

export default ExpenseListItem