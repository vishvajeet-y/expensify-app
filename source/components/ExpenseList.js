import React from 'react'
import {connect} from 'react-redux'
import selectExpenses from '../selectors/expenses'
import ExpenseListItem from './ExpenseList-item'
export const ExpenseList=(props)=>(
    <div>
    {
           
             props.expenses.length===0 ? <p>No Expenses</p>  : 
              props.expenses.map((expense)=>{
                return  <ExpenseListItem key={expense.id} {...expense}/>
             })
        
    }
   
    </div>
)

const mapStateToProps=(state)=>{
    return {
       expenses:selectExpenses(state.expenses,state.filters)
    }

}

export default connect(mapStateToProps)(ExpenseList)

// const ConnectedExpenseList=connect((state)=>{
//     return {
//        expenses:state.expenses,
//        filters:state.filters
//     }
// })(ExpenseList)

// export default ConnectedExpenseList