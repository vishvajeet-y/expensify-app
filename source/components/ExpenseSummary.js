import React from 'react'
import {connect} from 'react-redux'
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import ExpenseTotal from '../selectors/ExpenseTotal'

export const ExpenseSummary =({expensesCount,expensesTotal})=>{
       const expenseWord=expensesCount===1?'expense':'expenses'
      
       const formatedExpenseTotal=numeral(expensesTotal).format('$0,0.00')
        return <div>
       <h1>viewing {expensesCount} {expenseWord} totaling {formatedExpenseTotal}</h1>
        </div>

}
const mapToProps=(state)=>{
    const visibleExpenses=selectExpenses(state.expenses,state.filters)
    return {
            expensesCount:visibleExpenses.length,
            expensesTotal:ExpenseTotal(visibleExpenses)
    }
}

export default connect(mapToProps,)(ExpenseSummary)