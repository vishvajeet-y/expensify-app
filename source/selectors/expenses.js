import moment from 'moment'
const getVisibleExpenses=(expenses,{text,sortBy,startDate,endDate})=>{
    return expenses.filter((expense)=>{
        const createAtMoment=moment(expense.createAt)
        const startDateMatch=startDate?startDate.isSameOrBefore(createAtMoment) :true
        const endDateMatch=endDate ?endDate.isSameOrAfter(createAtMoment) :true
        const textMatch=expense.description.toLowerCase().includes(text.toLowerCase())
        return startDateMatch&&endDateMatch&&textMatch
    }).sort((a,b)=>{
        if(sortBy=='date')
        return a.createAt<b.createAt?1:-1
        if(sortBy=='ammount')
        return a.ammount<b.ammount ?1:-1
    })
   }
   export default getVisibleExpenses