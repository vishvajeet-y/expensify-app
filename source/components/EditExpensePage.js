import React from 'react'
import {connect} from 'react-redux'
import {editExpense,removeExpense} from '../actions/expenses'

import ExpenseForm from './ExpenseForm'

export class EditExpensePage extends React.Component {
   onSubmit=(expense)=>{
    this.props.editExpense(this.props.expense.id,expense)
    this.props.history.push('/')
   }

   onRemove=()=>{
      this.props.removeExpense(this.props.expense.id)
      this.props.history.push('/')
       
   }
  render(){

    return  <div>
    <ExpenseForm 
   expense={this.props.expense} onSubmit1={this.onSubmit}   />
    <button onClick={this.onRemove} >Remove</button>    
    </div>
  }
}

const mapStateToProps=(state,props)=>(

    {
      expense:state.expenses.find((expense)=>{
        return expense.id===props.match.params.id
      })
    }
  )

  const mapDispatchToProps =(dispatch)=>({
   editExpense:(id,expense)=>(dispatch(editExpense(id,expense))),
   removeExpense:(id)=>(dispatch(removeExpense({id})))
  })

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)