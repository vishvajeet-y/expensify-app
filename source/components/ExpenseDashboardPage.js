import React from 'react'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseList from './ExpenseList'
import ExpenseSummary from './ExpenseSummary'
const ExpenseDashboardPage=()=>(
    <div>
    <ExpenseSummary />
    <ExpenseListFilters/>
    <ExpenseList/>
    </div>
)
export default ExpenseDashboardPage