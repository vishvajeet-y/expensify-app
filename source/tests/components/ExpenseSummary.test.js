import React from 'react'
import {shallow} from 'enzyme'
import expenses from '../fixture/expenses'
import {ExpenseSummary} from '../../components/ExpenseSummary'
test('shouled correctly render ExperimentSummary one expense',()=>{
    const wrapper=shallow(<ExpenseSummary expensesCount={1} expensesTotal={123}/>) 
    expect(wrapper).toMatchSnapshot()
})

test('shouled correctly render ExperimentSummary with multiple expense',()=>{
    const wrapper=shallow(<ExpenseSummary expensesCount={23} expensesTotal={123343}/>) 
    expect(wrapper).toMatchSnapshot()
})