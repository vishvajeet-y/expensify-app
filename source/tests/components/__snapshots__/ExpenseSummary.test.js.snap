// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shouled correctly render ExperimentSummary one expense 1`] = `
<div>
  <h1>
    viewing 
    1
     
    expense
     totaling 
    $123.00
  </h1>
</div>
`;

exports[`shouled correctly render ExperimentSummary with multiple expense 1`] = `
<div>
  <h1>
    viewing 
    23
     
    expenses
     totaling 
    $123,343.00
  </h1>
</div>
`;
